archive/map-reduce.md

```diff
@@ -0,0 +1,116 @@
+# Map-reduce
+
+## Motivation
+
+There are hundreds of programming models to handle some data sets. But the Google has some problems to handle their data sets.
+
+- The input data is large.
+- The computations have to be distributed across more than hundreds of machines.
+
+To handle those data sets, the below detalis must be implemented.
+
+- how to parallelized the computation
+- how to distribute the data
+- how to handle failures
+
+After Google implemented hundreds of such programming models, they made an abstraction of them, that is Map-reduce model.
+
+## Model definition
+
+The programming model can be defined as the input and output of its execution. There are three execution model on Map-reduce model, that are the full program, map function and reduce function.
+
+- Full execution
+  - Input: a set of key/value pair
+  - Output: a set of key/value pair
+- Map: the function of producing intermediate key/value pair
+  - Input: a key/value pair
+  - Output: a set of key/value pair
+- Reduce: the function of producing intermediate key/value pair
+  - Input: an intermediate key I and a set of values for that key.
+  - Output: the desired ouput on the given intermediate key I
+With pseudocode, we can write as below. The specific implementation needs to be written by the developers.
+
+```
+def program(input: Pair<K, V>): Output<K, O>
+  do 
+
+def map(input: Pair<K, V>): Set<Intermediates<T, W>>
```

Comment:
```suggestion
def map(input: Pair<K, V>): Set<Intermediate<T, W>>
```

Reply:
I applied this at 217e370f276880aa93ef23c1d833c48e533d6191.